### PR TITLE
[ShapeOpt] Interface that allows running an optimization with SU2

### PR DESCRIPTION
**Pull request by dbaumgaertner**
This PR adds an interface that allows running an optimization with the open-source CFD code [SU2](https://su2code.github.io/).

An example can be found in the examples repository [here](https://github.com/KratosMultiphysics/Examples/tree/master/shape_optimization):

### CODE REVIEW

**Comment by philbucher** [comment]
Seems to be suitable also for CoSimulation ;)

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I think that the addition of this to the master branch should be considered in the @KratosMultiphysics/technical-committee (and possibly in the Kratos Steering Committee).

**Comment by dbaumgaertner** [comment]
@rubenzorrilla Just out of curiosity, why do you think it must be specifically considered by the technical and steering committee? I mean this little script is really tailored to optimization. But indeed, I should have given this PR a more descriptive title.

In any case, I am open for any feedback or advice in this context.

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla Just out of curiosity, why do you think it must be specifically considered by the technical and steering committee? I mean this little script is really tailored to optimization. But indeed, I should have given this PR a more descriptive title.
> 
> In any case, I am open for any feedback or advice in this context.

More than programming its kind of political issue. Do we want to open the possibility that our master branch links to other Open Source codes that have similar capabilities to our in-house ones? I think this should be at least discussed.

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+            sys.stdout = old_stdout
+
+# ==============================================================================
+class InterfaceSU2():
+
+    # --------------------------------------------------------------------------
+    def __init__(self,interface_parameters ):
+        default_parameters = KM.Parameters("""
+        {
+            "su2_related":
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:38]: `_related` reads strange, how about `su2_parameters` or `_settings`?

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+        self.element_id_su2_to_kratos = {}
+        self.list_of_optimization_conditions = []
+
+        self.__ReadSU2Mesh()
+        self.__TranslateSU2IdsToKratosIds()
+
+    # --------------------------------------------------------------------------
+    def WriteSU2MeshAsMDPA(self):
+        if self.interface_parameters["echo_level"].GetInt()>0:
+            print("> Start writing mdpa file...")
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:81]: please use the logger, as in the other scripts of ShapeOpt

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+        self.__WriteSubModelPartsWithoutDesignSurface()
+
+        self.new_file.close()
+
+        if self.interface_parameters["echo_level"].GetInt()>0:
+            print("> Finished writing mdpa file!\n")
+
+    # --------------------------------------------------------------------------
+    def WriteNodesAsSU2MeshMotionFile(self,kratos_node_set,target_directory = "."):
+        self.new_file = open(os.path.join(target_directory,self.interface_parameters["su2_related"]["mesh_motion_file"].GetString()), 'w')
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:101]: `new_file` is only used in this function, can be local variable. Also you could use the `with open() as new_file` pattern

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+
+        self.__ReadSU2Mesh()
+        self.__TranslateSU2IdsToKratosIds()
+
+    # --------------------------------------------------------------------------
+    def WriteSU2MeshAsMDPA(self):
+        if self.interface_parameters["echo_level"].GetInt()>0:
+            print("> Start writing mdpa file...")
+
+        self.new_file = open(self.interface_parameters["kratos_related"]["mdpa_file"].GetString(), 'w')
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:83]: please use the `with open(...) as ...` pattern

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+        kratos_gradient = self.__TranslateGradientToKratosFormat(su2_gradient)
+
+        self.project.config["WRT_SOL_FREQ"] = direct_solution_frequency
+        self.project.config["CONV_RESIDUAL_MINVAL"] = direct_residual_min
+        self.project.config["RESTART_SOL"] = direct_restart_option
+
+        return kratos_gradient
+
+    # --------------------------------------------------------------------------
+    def ReadNodalValuesFromCSVFile(self, filename, collum_with_node_id, collums_to_read, number_of_header_rows):
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:183]: typo: column

Also this method could be a function

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+
+        return values
+
+    # --------------------------------------------------------------------------
+    def __ReadSU2Mesh(self):
+
+        if self.interface_parameters["echo_level"].GetInt()>0:
+            print("\n> Start reading SU2 mesh file...")
+
+        ''' imports mesh and builds python dictionary structure
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:216]: docstring belongs directly under function name, not in the middle of the method

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+                keepon = False
+                break
+
+            # fix white space
+            line = line[0]
+            line = line.replace('\t',' ')
+            line = line.replace('\n',' ')
+
+            # skip comments
+            if line[0] == "%":
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:260]: Is it possible that there is a space before the `%`?

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+
+                # only read nelem lines
+                fileslice = islice(f_tb_read,nelem)
+
+                # the data pattern
+                pattern = tuple( [int] + [int]*9 )
+
+                # scan next lines for element data
+                elem = [
+                    [ t(s) for t,s in zip(pattern,line.split()) ]
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:284]: I don't know this syntax, out of curiosity: Does it do any kind of validation? 
* transforming values to ints
* checking for the correct number of entries?

Also `tuple( [int] + [int]*9 )`is the same as `tuple( [int]*10 )`

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+
+                # start SU2_MARK dictionary
+                thismark = {}
+
+                # read number of marker elements
+                line = mesh_readlines()[0]
+                if not "MARKER_ELEMS=" in line:
+                    raise Exception("Marker Specification Error")
+
+                # convert string to int int
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:339]:
```suggestion
                # convert string to int
```

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+                thismark["NELEM"] = thisnelem
+
+                # only read thisnelem lines
+                fileslice = islice(f_tb_read,thisnelem)
+
+                # the data pattern
+                pattern = tuple( [int] + [int]*9 )
+
+                # scan next lines for element data
+                markelem = [
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:352]: this same code for reading the elements is duplicated

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+
+        # Translate element ids
+        for elem in self.su2_mesh_data["ELEM"]:
+            su2_id = elem[-1]
+
+            if su2_id == 0:
+                kratos_id = self.su2_mesh_data["NELEM"]
+                self.element_id_su2_to_kratos[su2_id] = kratos_id
+            else:
+                kratos_id = elem[-1]
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:399]:
```suggestion
                kratos_id = su2_id
```

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 12]
```diff
@@ -0,0 +1,730 @@
...
+
+        condition_id = 0
+
+        # write line elements on design surface as line conditions
+        if line_entries:
+            self.new_file.write("Begin Conditions LineCondition2D2N\n")
+            for list_index in line_entries:
+                entry = self.su2_mesh_data["MARKS"][design_surface_tag]["ELEM"][list_index]
+
+                condition_id = condition_id+1
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:619]:
```suggestion
                condition_id += 1
```

File: applications/ShapeOptimizationApplication/python_scripts/interface_su2.py [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+                self.new_file.write("\n")
+            self.new_file.write("End Conditions\n\n")
+
+        # write triangular elements on design surface as triangle conditions
+        if tria_entries:
+            self.new_file.write("Begin Conditions SurfaceCondition3D3N\n")
+            for list_index in tria_entries:
+                entry = self.su2_mesh_data["MARKS"][design_surface_tag]["ELEM"][list_index]
+
+                condition_id = condition_id+1
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:639]:
```suggestion
                condition_id += 1
```

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/interface_su2.py:639]: same for the following

**Review by armingeiser — CHANGES_REQUESTED** [review]
In general it looks ok to me. I suggest putting it in a subfolder, to better separate it from "internal" code.

**Comment by adityaghantasala** [comment]
> More than programming its kind of political issue. Do we want to open the possibility that our master branch links to other Open Source codes that have similar capabilities to our in-house ones? I think this should be at least discussed.

Well Co-Sim app is already in Master and it is designed to do exactly that.  The point for me would be to have all the possibilities of communicating  with externally solvers/tools centrally at one place inside of Kratos. So that it is clean and well defined. 

In any case, one can already put this SU2 wrapper inside of CoSimApp and use it in ShapeOptimizationApp via CoSimmApp. 

I will also encourage @dbaumgaertner and others from @KratosMultiphysics/optimization   to consider this possibility of doing optimization via CoSimApp.

**Comment by rubenzorrilla** [comment]
> > More than programming its kind of political issue. Do we want to open the possibility that our master branch links to other Open Source codes that have similar capabilities to our in-house ones? I think this should be at least discussed.
> 
> Well Co-Sim app is already in Master and it is designed to do exactly that. The point for me would be to have all the possibilities of communicating with externally solvers/tools centrally at one place inside of Kratos. So that it is clean and well defined.
> 
> In any case, one can already put this SU2 wrapper inside of CoSimApp and use it in ShapeOptimizationApp via CoSimmApp.
> 
> I will also encourage @dbaumgaertner and others from @KratosMultiphysics/optimization to consider this possibility of doing optimization via CoSimApp.

Agree. Some days ago @philbucher and I reached the similar conclusion.

**Comment by dbaumgaertner** [comment]
Thanks all for the comments! Looking into the CoSimApp I totally agree that it fits much better in there.  I wasn't aware that the App is already such mature. In fact, features are great from what I see. So I will change this PR and instead, propose an SU2 wrapper in the CoSimApp (which still just interfaces optimization relevant things in SU2, though). I will come back to @philbucher or @adityaghantasala in this context.

As far as the realization of the optimization workflow with the Co-Sim App is concerned, I see the possibilities of setting up a process flow (co-simulation) with the app. Still, I am a bit more hesitating in this context, as I do not yet see an advantage in realizing the optimization algorithm within some json-settings, which requires a basic understanding of the Co-Sim workflow, rather than directly putting it in one single python script. Also, the output and the actual simulations are more tricky as there are many cross-dependencies. I am sure, this all could be realized with the Co-Sim App somehow. I just wonder what is the advantage of bringing in an additional level of complexity in the already confusing enough world of Vertex Morphing. Obviously, this is just a guts feeling and should be tested carefully. Maybe the implementation is far simpler than I can think of now and we could indeed use synergies. 

In any case, having and using the wrappers in the Co-Sim App is great for the optimization capabilities!

**Comment by philbucher** [comment]
I agree with @dbaumgaertner that this needs some figuring out of things!

In the long run the goal would be to reduce complexity and not add more. I think we can manage this by clearly separating the responsibilities: CoSimApp communicates with external solvers, ShapeOptApp uses this information then

Also I wanted to mention that the [CoSimIO](https://github.com/KratosMultiphysics/CoSimIO) which had its first release yesterday would be a good fit for this usecase

**Comment by adityaghantasala** [comment]
@philbucher I think  CoSimIO  would be an overkill in this case. Since SU2 has a good python interface a simple IO for SU2 should suffice. 

By the way congratulations on the first release of CoSimIO. !!

**Comment by philbucher** [comment]
> @philbucher I think CoSimIO would be an overkill in this case. Since SU2 has a good python interface a simple IO for SU2 should suffice.

I am not sure
Most of this interface deals with translating and reading/writing meshes and gradients from one to another format. This is essentially what the CoSimIO does too. 
I guess it would be possible to access SU2 directly and get the mesh (similarly to how the gradient is accessed). This mesh could be given to the CoSimIO which would do the transfer to/from Kratos
Same for the Gradient

**Review by rubenzorrilla — DISMISSED** [review]
@KratosMultiphysics/technical-committee also favors the coupling with other codes through the co-simulation application. Removing the blockages so you can proceed with the migration.

**Review by philbucher — COMMENTED** [review]
image

please add the big example to the Examples repo
also this PR must be squashed otherwise we have those huge files in the history! => [how to squash a PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request)

more detailed review coming in the next days

**Comment by dbaumgaertner** [comment]
Sorry, it seems that somehow I have required another review. This is not necessary. I am going to withdraw this PR and add the interface using the Co-Sim App if necessary. Regarding the examples. Since this PR is to be used by somebody less familiar with the Kratos structure I wanted to keep things simple and in one place. But obviously, this is not the best way. I am going to put the examples in the examples repo. Thanks, @philbucher for your hints!

**Comment by philbucher** [comment]
do you still need the branch?

**Comment by dbaumgaertner** [comment]
yes, for this year. As I said, somebody else still works with it. But I removed /squashed all memory-blasting commits. So apart from another entry in the list of branches, the footprint of this branch should be very low.

**Pull request closed without merging**